types/index.d.ts

```diff
@@ -0,0 +1,565 @@
+   *         "bpp": 32
+   *     }
+   * }
+   * ```
+   *
+   * @deprecated
+   */
+  get_statistics(): Object
+  get_instruction_counter(): number
+  is_running(): boolean
+
+  /**
+   * Send a sequence of scan codes to the emulated PS2 controller. A list of
+   * codes can be found at http://stanislavs.org/helppc/make_codes.html.
+   * Do nothing if there is no keyboard controller.
+   *
+   * @param codes
+   */
+  keyboard_send_scancodes(codes: number[]): void
+  
+  /**
+   * Send translated keys
+   */
+  keyboard_send_keys(codes: any[]): void
+
+  /**
+   * Send text
+   */
+  keyboard_send_text(string: string): void
+
+  
+  /**
+   * Download a screenshot.
+   */
+  screen_make_screenshot(): void
+
+  /**
+   * Set the scaling level of the emulated screen.
+   *
+   * @param {number} sx
+   * @param {number} sy
+   *
+   * @ignore
+   * @export
+   */
+  screen_set_scale(sx: number, sy: number): void
+
+  /**
+   * Go fullscreen.
+   */
+  screen_go_fullscreen(): void
+
+  /**
+   * Lock the mouse cursor: It becomes invisble and is not moved out of the browser window.
+   */
+  lock_mouse(): void
+
+  /**
+   * Enable or disable sending mouse events to the emulated PS2 controller.
+   */
+  mouse_set_status(enabled: boolean): void
+  
+  /**
+   * Enable or disable sending keyboard events to the emulated PS2 controller.
+   */
+  keyboard_set_status(enabled: boolean): void
+
+  /**
+   * Send a string to the first emulated serial terminal.
+   *
+   * @param data
+   */
+  serial0_send(data: string): void
+
+  /**
+   * Send bytes to a serial port (to be received by the emulated PC).
+   *
+   * @param serial the index of the serial port
+   * @param data
+   */
+  serial_send_bytes(serial: number, data: Uint8Array)
+
+  /**
+   * Mount another filesystem to the current filesystem.
+   * @param path Path for the mount point
+   * @param baseurl
+   * @param basefs As a JSON string
+   * @param callback
+   * @export
+   */
+  mount_fs(path: string, baseurl?: string, basefs?: string, callback?: (error: Object | null) => void): void
+
+  /**
+   * Write to a file in the 9p filesystem. Nothing happens if no filesystem has
+   * been initialized. First argument to the callback is an error object if
+   * something went wrong and null otherwise.
+   *
+   * @param file
+   * @param data
+   * @param callback
+   */
+  create_file(file: string, data: Uint8Array, callback?: (error: Object | null) => void): void
+
+  /**
+   * Runs a set of automatic steps
+   * @param steps
+   */
+  automatically(steps: V86AutomaticStep[]): void
+
+  /**
+   * Reads data from memory at specified offset.
+   *
+   * @param offset
+   * @param length
+   */
+  read_memory(offset: number, length: number): Array<number> | Uint8Array
+
+  /**
+   * Writes data to memory at specified offset.
+   *
+   * @param {Array.<number>|Uint8Array} blob
+   * @param {number} offset
+   */
+  write_memory(blob: Array<number> | Uint8Array, offset: number): void
+}
```

Comment:
This is brilliant, thanks! Did you write this from scratch or is (at least partially) generated from some of the source files? I'm a bit worried this might become out-of-date over time.

(nitpick: Add a newline at the end of the file)

Minor suggestion: Could you add a test, something very simple (e.g., similar to tests/api/clean-shutdown.js) that demonstrates loading this file with ts-node and using V86 from typescript? Doesn't need to be part of CI or mentioned in the readme yet.

Reply:
@copy: this is generated from your JSDoc @param types. My approach was to scan through your `starter.js` and add typings for members with `@exports`,  `write_memory` or `read_memory` had no `@exports` on them, but It felt essential. 

> Minor suggestion: Could you add a test, something very simple (e.g., similar to tests/api/clean-shutdown.js) that demonstrates loading this file with ts-node and using V86 from typescript? Doesn't need to be part of CI or mentioned in the readme yet.

sure.